knowledge_verificator/utils/configuration_parser.py

```diff
@@ -0,0 +1,129 @@
+            implementation of experiments on language models.
+        experiment_results (Path): Path to a directory, where results
+            should be saved.
+    """
+
+    learning_materials: Path
+    logging_level: str
+    mode: OperatingMode
+    experiment_implementation: Path
+    experiment_results: Path
+    production_mode: bool
+
+    def __init__(
+        self,
+        **kwargs,  # Records received from a YAML file.
+    ) -> None:
+        # Assign values to attributes.
+        for attribute, value in kwargs.items():
+            self.__setattr__(attribute, value)
+
+        # Convert to a proper datatypes.
+        self.learning_materials: Path = Path(kwargs['learning_materials'])
+        self.mode: OperatingMode = OperatingMode(kwargs['mode'])
+        self.experiment_implementation = Path(
+            kwargs['experiment_implementation']
+        )
+        self.experiment_results = Path(kwargs['experiment_results'])
+
+
+class ConfigurationParser:
+    """Class, which loads and parses a YAML configuration."""
+
+    def __init__(self, configuration_file: Path) -> None:
+        """Load a YAML configuration file."""
+        self._config_file: Path | None = None
+        self._config_data: dict = {}
+        self.load_configuration(configuration_file)
+
+    def _parse_configuration_file(self) -> None:
+        if self._config_file is None:
+            raise ValueError(
+                'Configuration file was not supplied and has value of None. '
+                'Provide a path to the configuration file before parsing it.'
+            )
+        with open(self._config_file, 'rt', encoding='utf-8') as fd:
+            self._config_data = yaml.safe_load(fd)
+
+    def parse_configuration(self) -> Configuration:
+        """
+        Parse the previously loaded YAML configuration file to an instance
+        of `Configuration`.
+
+        Returns:
+            Configuration: Parsed configuration of the application.
+        """
+        configuration_arguments: dict[str, Any] = {}
+        # Attributes of `Configuration` are YAML keys.
+        for option_name, _ in Configuration.__annotations__.items():
+            configuration_arguments[option_name] = self._config_data[
+                option_name
+            ]
+
+        # Feed key, value pairs to constructor.
+        return Configuration(**configuration_arguments)
+
+    def load_configuration(self, configuration_path: Path) -> None:
+        """
+        Load a configuration from a YAML file.
+
+        This method should be used only if you want to:
+        - either load a new configuration file,
+        - or reload configuration file after it was changed.
+
+        Args:
+            configuration_path (Path | str): Path to a YAML configuration file.
+
+        Raises:
+            FileNotFoundError: Raised if a configuration file was not found.
+        """
+        if not configuration_path.exists():
```

Comment:
Btw, lines (124), and (89) introduce race condition. After the path is checked that it points at something, there is time for bad actor to delete it, and rise OSError when `open` is called

Reply:
So just `open` the thisg under provided path, and if everything is good, ok, if not, path was malformed or access to file was denied.

Reply:
You are right. `with open(...)` will emit its own error message anyway. Thank you!